README.md
# Monsterforge

## Features

You can find an overview and how to use video on the website.

* Create, upload and modify your creatures
* Import a [Dndbeyond encounter](https://www.dndbeyond.com/encounter-builder) into a bestiary, ready to print!
* Gather the creatures you need for a session in a bestiary
* Print minis for all the creatures in your bestiaries with a single click
* The Base color and shape can be customized
* Miniatures of the same type can be numbered
* The minis have a foldable base that can be reinforced with a coin to make them wind-resistant
* The optional name label on the miniature is sized to use all the space optimally
* The miniatures can be printed for 1-inch, 3/4-inch or 1/2-inch grids
* The height of the minis can be dynamically adjusted, based on creature size and source image
* You can adjust the brightness of the mini's backside to create a silhouette effect
* The minis are nicely arranged on a sheet in a way to minimize wasted space

## About the project

The website at [forge.dice.quest](https://forge.dice.quest) provides tools for tabletop gaming.
The website is free and contains no ads.
Feel free to contribute anything from commenting to design or new options and tools. If you plan something bigger, please contact me beforehand.

This project is supported by [JetBrains - PyCharm](https://www.jetbrains.com/pycharm/)!

### Fork

I took over as maintainer for the project originally made by matnad. I have since made it more perfomand, slimmer as well as added new features to it.
The original project is at [github.com/matnad/monsterforge](https://github.com/matnad/monsterforge), thanks to matnad for the initial work!

## Collaboration

If you wish to collaborate with me on anything, please contact me here.
I am open to any type of collaboration: Development, Graphics artist for website design or even for creature images, UI/UX, hosting (is currently fine), idea sharing, etc!

## Create and install virtual environment

```x-sh
git clone https://github.com/schemen/monsterforge.git && cd monsterforge
virtualenv venv -p python3
source venv/bin/activate
pip install -r requirements.txt
deactivate
```


CHANGELOG.md
# Changelog

## 1.11.0(2023-07-25)

### Improvement
* Implement Caching and Cache Busting


## 1.10.1(2023-07-19)

### Updates
* Dependency updates and maintenance

## 1.10.0(2023-02-28)

### Feature
* Add 28mm option as base (907e4e3 @indero)

### Updates
* Update dependencies

## 1.9.2(2022-12-23)

### Fix
* Fix an issue with the fake user agent
* Update some dependencies

## 1.9.1(2022-10-19)

### Improvements
* Dependency updates
* Switch to Python 3.10

## 1.9.0(2022-08-31)

### New
* Cavalry mode. This mode prints the creature on a rectangular base. Mostly for war-gaming.
* Temporary accounts cannot be created anymore. Existing ones can be converted as long as the session exists.

## 1.8.2(2022-09-03)

### Hotfix
* PDF size is now handled correctly, including a print margin
* Update django


## 1.8.1(2022-05-18)

### Hotfix
* Fix dtype issues when generating images

## 1.8.0(2022-05-18)

### New
* Add the possibility to change the creature background color. Default remains white.
* Added color selection tools to the classic builder as well as the Quickbuilder

### Improvements
* Transparent PNGs have their transparency replaced with the chosen background color.
* Applied a fix where enumeration doesn't apply correctly in the Quickbuilder and sometimes in the classic builder

## 1.7.0(2022-05-18)

### New
* Rearrange how images and nameboxes are sorted on a mini significantly increasing creature image size
  * Before, the namebox size has always been substracted from max_height in any configuration. Now, it will use that space for the image to maximise the size. If a namebox is requested, it will use available space first (max_height) and if not, it'll shrink the image to fit. Now, all space is optimally used and the minis are standardized in size.

## 1.6.5(2022-05-18)

### Improvements
* Image size and creature size are bigger now.
* If you choose a bigger paper size (A4 or Tabloid), Large, Huge and Gargantuan Minis are full height sized. Width is always correct of course.
* Improved the handling of creature position for cleaner minis and better size adjustments (Walking, Hovering, Flying)

## 1.6.4(2022-05-16)

### Improvements
* Added initial wave of unit tests
* Improved error handling, messages

## 1.6.3(2022-05-12)

### Improvements
* Improved english
* Improved descriptions
* Decreased docker image size by 200MB

## 1.6.2(2022-05-11)

### New
* Quickbuilder video added

### Improvements
* Small text tweaks and improvements

## 1.6.1(2022-05-11)

### Hotfix
* Hotfix for the classic minibuilder is released

## 1.6.0(2022-05-11)

### New
* The Quickbuilder beta is released!
  * Since this does not require an account, temporary accounts are deprecated and will be disabled in the future
  * If you want to keep your collections, make sure you convert to a real account

## 1.5.0(2022-05-04)

### New
* Instead of a ZIP file containing images, Monsterforge generates a PDF to print directly

### Improvements
* English has been improved

## 1.4.7(2022-05-04)

### Improvements
* Improve the /minis/ redirect
* Add robots.txt to protect authenticated views

## 1.4.6(2022-05-03)

### Hotfix
* More CR/CreatureType hotfixes

## 1.4.5(2022-05-03)

### Hotfix
* Fix the creature form

## 1.4.4(2022-05-03)

### Improvements
* Change some descriptions
* Smaller UI improvements
* Prevent signed in users from signup and temp account creation
* Remove D&D Agnostic data from creatures, this tool serves all
  * Create CR is no longer available as setting
  * Creature Type is no longer available as setting
  * This allows Monsterforge to be slimmer

## 1.4.3(2022-05-03)

### Improvements
* More descriptive title and meta description


## 1.4.2(2022-05-03)

### Improvements
* Redirect "minis/" to / for people that still the old url somewhere

## 1.4.1(2022-05-03)

### Improvements
* Remove "minis/" sub-url for cleaner urls

## 1.4.0(2022-04-29)

### Features
* First iteration of account management via profile page
  * You can change your password
  * You can delete your account which will delete all of your data

## 1.3.6(2022-04-25)

### Improvements
* Bump Django to fix CVE #55

## 1.3.5(2022-03-16)

### Improvements
* Update several dependencies

## 1.3.4(2022-01-13)

### Improvements
* Bump Django to fix CVE #51

## 1.3.3(2022-01-13)

### Improvements
* Update several dependencies


## 1.3.2(2021-09-22)

### Improvements
* Added Plausible Insights (Default=False)

## 1.3.1(2021-09-21)

### Improvements (So pretty!)
* Added new images on the Index page that better show the results
* Update sidebar image to better represent a forge

## 1.3.0(2021-09-20)

### Improvements
* Fixed many bugs with the DDB Importer (See #41)
* DDB Importer can now handle homebrewed monsters
* Log DDB URLs for better debugging

## 1.2.0 (2021-09-14)

### Improvements

* Code refactoring and cleanup
* Migrate to Django 3.2 (From Django 2.2)
* Enhance RPG Agnostic aspect 
* Several bugfixes
* Dependency updates

## 1.1.1 (2021-08-19)

### Improvements

* Circular minis now have a rectangular guiding lines, this will help crafting it in the end. Cut the Mini and glue it then cut the circle out.

### Other

* Updated the Readme
* Monsterforge is now it's own root

## 1.1.0 (2021-08-18)

### Improvements

* Refactored how images are processed and served, reducing server overhead
* Implemented some basic image caching for each requested build
* Add Link to Changelog notes in footer

## 1.0.7 (2021-08-17)

### fixes

* Fixed a problem where printing Names resulted in a 500 Error (#31)

## 1.0.6 (2021-08-05)

### fixes

* Significantly reduce image size
* Use github actions and registry
* Fix DDBs wrongly defined SRD monster image URL

## 1.0.5 (2021-08-05)

### features


paperminis/forms.py
import os
from urllib.parse import urlparse
from django import forms
from django.core.exceptions import ValidationError
from django.utils.translation import ugettext_lazy as _
from django.utils.safestring import mark_safe
from constrainedfilefield.fields import ConstrainedFileField
from django.conf import settings
from django.template.defaultfilters import filesizeformat
from django.core.exceptions import ValidationError
from .models import Creature, Bestiary, CreatureQuantity, PrintSettings

from django.contrib.auth.forms import UserCreationForm
from django.contrib.auth import get_user_model

User = get_user_model()


class UserDeleteForm(forms.Form):

    error_messages = {
        'password_incorrect': _("Your password was entered incorrectly. Please enter it again."),
    }

    def __init__(self, *args, **kwargs):
        self.user = kwargs.pop('user')
        super().__init__(*args, **kwargs)

    password = forms.CharField(
        label=_("Password"),
        strip=False,
        widget=forms.PasswordInput(attrs={'autocomplete': 'current-password', 'autofocus': True}),
    )

    def clean_password(self):
        """
        Validate that the password field is correct.
        """
        password = self.cleaned_data["password"]
        if not self.user.check_password(password):
            raise ValidationError(
                self.error_messages['password_incorrect'],
                code='password_incorrect',
            )
        return password

    class Meta:
        fields = ['password']

class SignUpForm(UserCreationForm):
    """Used to register new users"""
    # email
    email = forms.EmailField(max_length=254, help_text='Required. Inform a valid email address.')

    class Meta:
        model = User
        fields = ('email', 'password1', 'password2',)


class CreatureModifyForm(forms.ModelForm):
    """Add or Update creatures"""
    name = forms.CharField(widget=forms.TextInput(attrs={'class': 'form-control', 'size': '60%'}))

    class Meta:
        model = Creature
        fields = ['name', 'show_name', 'img_url', 'size', 'position', 'background_color', 'color', 'cavalry_mode']

    def __init__(self, *args, **kwargs):
        self.user = kwargs.pop('user')  # To get request.user. Do not use kwargs.pop('user', None) due to potential security hole
        # self.method = kwargs.pop('method') # To get create or update
        super(CreatureModifyForm, self).__init__(*args, **kwargs)
        self.fields['color'].required = False
        self.fields['background_color'].required = False
        self.fields['position'].required = False

    def clean(self):
        cleaned_data = super(CreatureModifyForm, self).clean()
        if self.user:
            name = cleaned_data.get("name")
            img_url = cleaned_data.get("img_url")
            # name + img_url must be unique. Otherwise raise a ValidationError
            count = Creature.objects.filter(owner=self.user, name=name, img_url=img_url).exclude(
                id=self.instance.id).count()
            if count == 1:
                raise forms.ValidationError(('This creature already exists. Use a different name or image url.'),
                                            code='exists', )

            # patreon early access backend validation
            # if self.user.groups.filter(name='Patrons').count() <= 0:
            #     cleaned_data['show_name'] = True
            #     cleaned_data['position'] = Creature.WALKING

        return cleaned_data


class BestiaryModifyForm(forms.ModelForm):
    """Simple bestiary create/update form"""
    name = forms.CharField(widget=forms.TextInput(attrs={'class': 'form-control', 'size': '80%'}))

    class Meta:
        model = Bestiary
        fields = ['name', ]


def validate_ddbenc_url(value):
    if not value:
        return  # Required error is done the field
    obj = urlparse(value)
    if not 'dndbeyond.com' in obj.netloc or not ('encounters/') in obj.path:
        raise ValidationError('Only DDB Encounter URLS are allowed!')


class DDBEncounterBestiaryCreate(forms.Form):
    """create bestiary and monsters from ddb encounter"""
    # URL
    ddb_enc_url = forms.URLField(help_text="Required. Enter a Dndbeyond Encounter URL.",
                                 validators=[validate_ddbenc_url])


class QuantityForm(forms.ModelForm):
    """Form to link creatures to bestiary. Most of this is done in views. """
    quantity = forms.IntegerField()

    # name = forms.CharField(max_length=150)
    class Meta:
        model = Creature
        fields = ['name', ]


class PrintForm(forms.ModelForm):
    """Simple Form for print settings."""

    class Meta:
        model = PrintSettings
        exclude = ['user']

    def clean_darken(self):
        darken = self.cleaned_data['darken']
        if not 0 <= darken <= 100:
            raise forms.ValidationError("Enter a value between 0 and 100")

        if not darken % 1 == 0:
            raise forms.ValidationError("Enter a multiple of 1")
        return darken


def validate_file_extension(value):
    """Function to validate json file extension and size."""
    ext = os.path.splitext(value.name)[1]
    valid_extensions = ['.json']
    if not ext.lower() in valid_extensions:
        raise ValidationError('Unsupported file extension. Please only upload .json files.')
    if value.size > settings.MAX_UPLOAD_SIZE:
        raise forms.ValidationError(_('Please keep filesize under %s. Current filesize %s') % (
        filesizeformat(settings.MAX_UPLOAD_SIZE), filesizeformat(value._size)))


class UploadFileForm(forms.Form):
    """Upload form with validator."""
    file = forms.FileField(validators=[validate_file_extension])


class QuickCreateSettingsForm(forms.Form):
    # paper format
    A4 = 'a4'
    A3 = 'a3'
    LETTER = 'letter'
    LEGAL = 'legal'
    TABLOID = 'tabloid'
    PAPER_FORMAT_CHOICES = (
        (A4, 'A4'),
        (A3, 'A3'),
        (LETTER, 'Letter'),
        (LEGAL, 'Legal'),
        (TABLOID, 'Tabloid')
    )
    # grid size
    GRID28 = 28
    GRID24 = 24
    GRID18 = 18
    GRID12 = 12
    GRID_SIZE_CHOICES = (
        (GRID28, '28 mm ~ 1.1 inch'),
        (GRID24, '24 mm ~ 1 inch'),
        (GRID18, '18 mm ~ 3/4 inch'),
        (GRID12, '12 mm ~ 1/2 inch')
    )
    # base shape
    SQUARE = 'square'
    HEXAGONAL = 'hexagon'
    CIRCLE = 'circle'
    BASE_SHAPE_CHOICES = (
        (SQUARE, 'Square'),
        (HEXAGONAL, 'Hexagon'),
        (CIRCLE, 'Circle')
    )

    # Print Settings
    paper_format = forms.ChoiceField(choices=PAPER_FORMAT_CHOICES, required=True)
    grid_size = forms.ChoiceField(choices=GRID_SIZE_CHOICES, required=True, initial=GRID24)
    base_shape = forms.ChoiceField(choices=BASE_SHAPE_CHOICES, required=True)
    enumerate = forms.BooleanField(required=False)


class QuickCreateCreatureForm(forms.Form):
    # Size
    TINY = 'T'
    SMALL = 'S'
    MEDIUM = 'M'
    LARGE = 'L'
    HUGE = 'H'
    GARGANTUAN = 'G'

    CREATURE_SIZE_CHOICES = (
        (TINY, 'Tiny'),
        (SMALL, 'Small'),
        (MEDIUM, 'Medium'),
        (LARGE, 'Large'),
        (HUGE, 'Huge'),
        (GARGANTUAN, 'Gargantuan')
    )

    # Base Color
    GREEN = '228b22'
    RED = 'aa3939'
    BLUE = '005b96'
    LIGHTGRAY = 'd3d3d3'
    DARKGRAY = 'a9a9a9'
    BLACK = '000000'
    WHITE = 'ffffff'

    COLOR_CHOICES = (
        (GREEN, 'Green'),
        (RED, 'Red'),
        (BLUE, 'Blue'),
        (LIGHTGRAY, 'Light Gray'),
        (DARKGRAY, 'Dark Gray'),
        (BLACK, 'Black'),
        (WHITE, 'White')
    )

    # Alignment
    FLYING = 'top'
    HOVERING = 'center'
    WALKING = 'bottom'

    POSITION_CHOICES = (
        (FLYING, 'Flying (Top)'),
        (HOVERING, 'Hovering (Middle)'),
        (WALKING, 'Walking (Bottom)')
    )
    # Creature Data
    img_url = forms.URLField(label='Image direct URL', required=False)
    position = forms.ChoiceField(choices=POSITION_CHOICES, initial="bottom")
    name = forms.CharField(max_length=100, required=False)
    size = forms.ChoiceField(choices=CREATURE_SIZE_CHOICES, initial="M")
    quantity = forms.IntegerField(max_value=100, min_value=1, required=False, initial=1)
    color = forms.ChoiceField(choices=COLOR_CHOICES, initial=DARKGRAY)
    background_color = forms.ChoiceField(choices=COLOR_CHOICES, initial=WHITE)
    cavalry_mode = forms.BooleanField(required=False)
